tests/integration/expected/fib.hir

```diff
@@ -8,13 +8,29 @@ global external @gv2 : i32 = $0 { id = 2 };
 
 pub fn fib(i32) -> i32 {
 block0(v0: i32):
-    v2 = const.i32 1000 : i32;
-    v3 = cast v0 : u32;
-    v4 = cast v2 : u32;
-    v5 = lt v3, v4 : i1;
-    v6 = cast v5 : i32;
-    br block1(v6);
+    br block2(v0);
 
 block1(v1: i32):
-    ret v1;
+
+block2(v3: i32):
+    v4 = const.i32 3 : i32;
+    v5 = cast v3 : u32;
+    v6 = cast v4 : u32;
+    v7 = gt v5, v6 : i1;
+    v8 = cast v7 : i32;
+    v9 = neq v8, 0 : i1;
+    condbr v9, block4, block5;
+
+block3(v2: i32):
```

Comment:
Is there any way to avoid emitting these useless blocks? They don't seem to be referenced by anything, and they contain no instructions, which seem like something we should be able to catch quite easily. Right now IR like this will fail our validator pass because we assert that there are no empty blocks (dead blocks are fine, but empty ones are technically invalid, especially if they are present in the CFG).

Reply:
I don't see an easy way of doing this in the frontend.
The `block1` is the "exit" block created right after the entry block. We branch there on the `end` op that represents the end of the function. In this code example, we always exit the function via the "early" `return` op. 
The `block3` is the "next" block after the loop block created on the "loop" op. We branch to it when we exit the loop "normally" by not branching on the next iteration. In this code example, we always branch on the next iteration and never exit the loop "normally" but only via `return`.

Can we use a special pass to eliminate such blocks that the frontend would call before returning an IR? OTOH, a DCE pass should be able to eliminate them.

Reply:
One trivial way to handle this, if a more precise method isn't viable, is to do a simple linear scan of all blocks in the function at the end of translating Wasm to the IR, and remove any blocks which are empty. Since it appears that in the Wasm translation, an empty block only appears when it is unreachable, it should be safe to remove such blocks without computing the CFG/checking for reachability. 

Empty blocks are not valid in the IR, so writing a generic pass to remove such blocks doesn't make much sense, but even if we were to do so, such a pass would require a CFG analysis to determine whether that block is reachable or not (because a reachable empty block is definitely an error). That's more expensive to compute, and if we can avoid that overhead, all the better. In practice, the `InlineBlocks` pass does this kind of thing already, but all of our analyses/passes assume that the IR has basic validity properties upheld (blocks aren't empty, and have a valid terminator, and that uses of SSA values are dominated by their definition).

Reply:
Good idea! I was way too fixated on the strictness of the "empty blocks not being a successor to any block" rule. I made https://github.com/0xPolygonMiden/compiler/issues/63 and will do it in my next PR if you don't mind.

Reply:
Sounds good!